nexus/src/app/metrics.rs

```diff
@@ -164,4 +165,82 @@ impl super::Nexus {
                 _ => Error::InternalError { internal_message: e.to_string() },
             })
     }
+
+    /// Run an OxQL query against the timeseries database, scoped to a specific project.
+    pub(crate) async fn project_timeseries_query(
+        &self,
+        _opctx: &OpContext,
+        project_lookup: &lookup::Project<'_>,
+        query: impl AsRef<str>,
+    ) -> Result<Vec<oxql_types::Table>, Error> {
+        // Ensure the user has read access to the project
+        let (authz_silo, authz_project) =
+            project_lookup.lookup_for(authz::Action::Read).await?;
+
+        // Ensure the query only refers to the project
+        let filtered_query = format!(
+            "{} | filter silo_id == \"{}\" && project_id == \"{}\"",
+            query.as_ref(),
+            authz_silo.id(),
+            authz_project.id()
+        );
+
+        self.timeseries_client
+            .oxql_query(filtered_query)
+            .await
+            .map(|result| result.tables)
+            .map_err(|e| match e {
+                oximeter_db::Error::DatabaseUnavailable(_) => {
+                    Error::ServiceUnavailable {
+                        internal_message: e.to_string(),
+                    }
+                }
+                oximeter_db::Error::Oxql(_)
+                | oximeter_db::Error::TimeseriesNotFound(_) => {
+                    Error::invalid_request(e.to_string())
+                }
+                _ => Error::InternalError { internal_message: e.to_string() },
+            })
+    }
+
+    /// List available project-scoped timeseries schema
+    pub(crate) async fn project_timeseries_schema_list(
+        &self,
+        opctx: &OpContext,
+        pagination: &TimeseriesSchemaPaginationParams,
+        limit: NonZeroU32,
+    ) -> Result<dropshot::ResultsPage<TimeseriesSchema>, Error> {
+        // any authenticated user should be able to do this
+        let authz_silo = opctx
+            .authn
+            .silo_required()
+            .internal_context("listing project-scoped timeseries schemas")?;
+        opctx.authorize(authz::Action::ListChildren, &authz_silo).await?;
+
+        self.timeseries_client
+            .timeseries_schema_list(&pagination.page, limit)
+            .await
+            .and_then(|schemas| {
+                let filtered = schemas
+                    .items
+                    .into_iter()
+                    .filter(|schema| schema.authz_scope == AuthzScope::Project)
+                    .collect();
+                dropshot::ResultsPage::new(
+                    filtered,
+                    &dropshot::EmptyScanParams {},
+                    |schema, _| schema.timeseries_name.clone(),
+                )
+                .map_err(|err| oximeter_db::Error::Database(err.to_string()))
+            })
```

Comment:
Just realized filtering here makes no sense because `limit` is being applied before filtering. We need a new method or a new arg in the oximeter client.